tests/unit/core.js

```diff
@@ -183,24 +183,109 @@ exports.testNewNonNativeObject = function (test) {
   test.done();
 };
 
+exports["undefined as variable name"] = function (test) {
 
-/**
- * Test that JSHint allows `undefined` to be a function parameter.
- * It is a common pattern to protect against the case when somebody
- * overwrites undefined. It also helps with minification.
- *
- * More info: https://gist.github.com/315916
- */
-exports.testUndefinedAsParam = function (test) {
-  var code  = '(function (undefined) {}());';
-  var code1 = 'var undefined = 1;';
+  TestRun(test)
+    .test("var undefined;");
 
-  TestRun(test).test(code);
+  TestRun(test)
+    .test("let undefined;", { esnext: true });
+
+  TestRun(test)
+    .addError(1, "const 'undefined' is initialized to 'undefined'.")
+    .test("const undefined;", { esnext: true });
+
+  TestRun(test)
+    .test("var { undefined } = {};", { esnext: true });
+
+  TestRun(test)
+    .test("var [ undefined ] = [];", { esnext: true });
+
+  var functions = [
+    "function a(undefined) {}",
+    "function b({ undefined }) {}",
```

Comment:
I am not sure I like allowing undefined to be in destructuring expressions.
It *is* technically valid, you are right, but from a jshint perspective its confusing - we cannot guarentee that people will not pass `{ undefined: true }` and end up with what we were trying to protect against.
I think the patterns of having it as an argument and having it as a un-assigned variable is enough freedom ?